src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -279,20 +301,31 @@ internal static void SetProperty(object resource, string propertyName, object va
             }
         }
 
-        internal static void SetInstanceAnnotations(ODataResource oDataResource, IODataInstanceAnnotationContainer instanceAnnotationContainer, 
-            ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
+        internal static void SetInstanceAnnotations(
+            ODataResourceBase oDataResource,
+            IODataInstanceAnnotationContainer instanceAnnotationContainer,
+            IODataInstanceAnnotationContainer transientAnnotationContainer,
+            ODataDeserializerProvider deserializerProvider,
+            ODataDeserializerContext readContext)
         {
-            if(oDataResource.InstanceAnnotations != null)
+            if (oDataResource.InstanceAnnotations != null)
             {
                 foreach (ODataInstanceAnnotation annotation in oDataResource.InstanceAnnotations)
                 {
-                    AddInstanceAnnotationToContainer(instanceAnnotationContainer, deserializerProvider, readContext, annotation,string.Empty);
+                    if (!TransientAnnotations.TransientAnnotationTerms.Contains(annotation.Name))
+                    {
+                        AddInstanceAnnotationToContainer(instanceAnnotationContainer, deserializerProvider, readContext, annotation, string.Empty);
+                    }
+                    else
+                    {
+                        AddInstanceAnnotationToContainer(transientAnnotationContainer, deserializerProvider, readContext, annotation, string.Empty);
+                    }
                 }
             }
 
-            foreach(ODataProperty property in oDataResource.Properties)
+            foreach (ODataProperty property in oDataResource.Properties)
             {
-                if(property.InstanceAnnotations != null)
+                if (property.InstanceAnnotations != null)
```

Comment:
This check is superfluous. The `InstanceAnnoations` property cannot return null.
See
1) https://github.com/OData/odata.net/blob/b37a90e32319636c368a2ce9e5ba2cf39cb65582/src/Microsoft.OData.Core/ODataPropertyInfo.cs#L38
2) https://github.com/OData/odata.net/blob/b37a90e32319636c368a2ce9e5ba2cf39cb65582/src/Microsoft.OData.Core/ODataAnnotatable.cs#L33